### DISCUSSION TITLE
Option to approve new comments

### DISCUSSION THREAD

**Discussion by retifrav** [umputun/remark42#1019] [Ideas]
That would be especially useful in case if anonymous authentication is enabled.

So the suggestion is to have an option that would enable a kind of "pre-approval" mode, in which a new comment from anonymous user (*or perhaps also the first ever comment from a user who just signed-in with either of social media*) would not be visible to other users until administrator approves it.

Once approved, the comment will become visible, and all further comments from that user won't need to go through pre-approval.

**Comment by umputun** [comment]
this is an interesting idea and can be useful in some cases. 

the part "Once approved, the comment will become visible, and all further comments from that user won't need to go through pre-approval." is likely for the regular users only because for anonymous this won't be possible to do.

**Reply by paskal** [reply]
Was writing this as well. Would be impossible to approve anonymous for a particular name because anyone can pick any anonymous name: so technically it's only possible to do pre-approval for "the first comment for a new user, or any comment from anonymous".

**Reply by retifrav** [reply]
Yep, didn't phrase that clearly, this "once-approved" part would indeed work only for "proper" signed-in users, and anonymous comments would always go through pre-approval.

**Reply by abow** [reply]
I'd also like the feature to always approve anonymous comments.  Thanks

**Comment by umputun** [comment]
Technically this could be done by introducing a `Pending` flag to the `Comment` struct and some "Approve" method to both store (service) and REST. From this point, we can show such comments in the admin-only view as "pending" with a link/button to approve. Probably should be not that hard on UI side (@akellbl4 am I right?)

All the other parts like "preapproved" logic most likely can be achieved with the combination of existing calls and the new pending flag. I.e. create comment method won't set a pending flag for any user with any existing comments.

_note: the reason why I want to go with the pending flag instead of approved is to keep it compatible with the current comments._

**Reply by akellbl4** [reply]
Looks pretty easy to implement at first glance.

**Reply by paskal** [reply]
Copying @umputun response from #1236 to preserve all the information in this discussion:

> There are two ways to implement it. One possibility is probably easy to do - if this pre-moderation flag is set we publish each new comment with a soft delete initially and latter on admin may "undelete" it. As far as I recall we already have API to do it. This solution can cause some unexpected complexity, for example, notification should be sent to the admin "channel" right away but postponed till approval before it is sent to users. Another problem with this solution - it is some kind of hack and `delete` status is not a proper substitution for the "enabled" (or "approved") flag
> 
> The proper solution is likely harder to implement, i.e. add `enabled` field. The backend change to support this new flag will require several changes in multiple places, for instance, need a way to get only those with enabled=false, change it to true, exclude comments from the regular (non-admin) view prior to approval, and so on. It will also need UI change, to approve/disapprove

**Comment by nununo** [comment]
Hello,

Any news on this? I also would like to allow for anonymous comments but I wouldn't do it unless these get to be pre-approved.

Thanks!

**Reply by paskal** [reply]
I've created #1300 about it and will look into it after solving some other issues in open-source projects from my todo, ETA might be somewhere within the next three months (if there would be no other virus or war). But please note that it would require front-end support as well, and we are historically short on hands for the front-end part.

**Reply by abow** [reply]
I'm willing to implement the front-end part when the time comes.

**Comment by cowsay1** [comment]
In remark42 I'm really missing fully featured moderation ui like disqus.com/admin/moderate
where I could review/manage/edit all comments and users without browsing every page of my blog

**Reply by geerlingguy** [reply]
I agree, it would be nice to have a little admin frontend; you can kinda emulate it with a custom page that pulls in latest comments, but it's a bit clunky. But that feature might be slightly different than this request (even if this would be part of it).

**Comment by Nicole4Fox** [comment]
I think this feature would be quite useful to have especially when allowing anonymous comments.

**Reply by geerlingguy** [reply]
Agreed; it's how I've managed to keep a lid on trolls/spammers on my own site—before I had comments publish automatically, and I would sometimes get someone who pops through 50+ blog posts and comments things (either spam links, or dummy text like "test"), and cleaning that all up is a nightmare. If they can't impact the public-facing website immediately, they tend to give up after one or two attempts.
